sw/device/silicon_creator/lib/drivers/kmac.c

```diff
@@ -0,0 +1,279 @@
+  // Set `CFG.STATE_ENDIANNESS` bit to 0 (little-endian).
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_STATE_ENDIANNESS_BIT, 0);
+  // Set `CFG.SIDELOAD` bit to 0 (no sideloading).
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_SIDELOAD_BIT, 0);
+  // Set `CFG.ENTROPY_MODE` field to use software entropy. SHAKE does not
+  // require any entropy, so there is no reason we should wait for entropy
+  // availability before we start hashing.
+  cfg_reg =
+      bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_MODE_FIELD,
+                             KMAC_CFG_SHADOWED_ENTROPY_MODE_VALUE_SW_MODE);
+  // Set `CFG.ENTROPY_FAST_PROCESS` bit to 0.
+  cfg_reg = bitfield_bit32_write(cfg_reg,
+                                 KMAC_CFG_SHADOWED_ENTROPY_FAST_PROCESS_BIT, 0);
+  // Set `CFG.MSG_MASK` bit to 0.
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_MSG_MASK_BIT, 0);
+  // Set `CFG.ENTROPY_READY` bit to 1.
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_READY_BIT, 1);
+  // Set `CFG.ERR_PROCESSED` bit to 0.
+  cfg_reg =
+      bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_ERR_PROCESSED_BIT, 0);
+  // Set `CFG.EN_UNSUPPORTED_MODESTRENGTH` bit to 0.
+  cfg_reg = bitfield_bit32_write(
+      cfg_reg, KMAC_CFG_SHADOWED_EN_UNSUPPORTED_MODESTRENGTH_BIT, 0);
+  abs_mmio_write32_shadowed(kBase + KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+
+  // Write entropy seed registers (all-zero).
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_0_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_1_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_2_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_3_REG_OFFSET, 0);
+  abs_mmio_write32(kBase + KMAC_ENTROPY_SEED_4_REG_OFFSET, 0);
+
+  // Read back and double-check the security strength and mode.
+  cfg_reg = abs_mmio_read32(kBase + KMAC_CFG_SHADOWED_REG_OFFSET);
+  HARDENED_CHECK_EQ(
+      bitfield_field32_read(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD),
+      KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L256);
+  HARDENED_CHECK_EQ(
+      bitfield_field32_read(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD),
+      KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE);
+
+  return kErrorOk;
+}
+
+rom_error_t kmac_shake256_start(void) {
+  // Block until KMAC hardware is idle.
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_IDLE_BIT));
+
+  // Issue `CMD.START` to start the operation.
+  issue_command(KMAC_CMD_CMD_VALUE_START);
+
+  // Block until KMAC hardware is in the `absorb` state. After `CMD.START`,
+  // KMAC should never move out of the `absorb` state until `CMD.PROCESS` is
+  // issued, so we get significant performance gains by polling only once here
+  // instead of before every `absorb`.
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_ABSORB_BIT));
+
+  return kErrorOk;
+}
+
+inline void kmac_shake256_absorb(const uint8_t *in, size_t inlen) {
```

Comment:
All the core SHAKE256 operations are performance-critical for SPHINCS+, so I marked some short ones `inline` to avoid the function-call overhead while I was optimizing the code. But based on some quick performance tests, it makes no significant difference, so I'll remove the `inline`.